File: haskell/74_Digit_factorial_chains.hs
import Data.Array.Unboxed
import Data.Array.Base (unsafeAt)

facts :: UArray Int Int
facts = array (0,9) [(0,1), (1,1), (2,2), (3,6), (4,24), (5,120), (6,720), (7,5040), (8,40320), (9,362880)]

factorial :: Int -> Int
factorial = unsafeAt facts

digitFactorialSum :: Int -> Int
digitFactorialSum = go 0
  where go acc n = case n `divMod` 10 of
                     (0,r) -> acc + factorial r
                     (q,r) -> go (acc + factorial r) q

chainLength :: Int -> Int
chainLength = go 0
  where go acc n = case n of
                     1 -> acc + 1
                     2 -> acc + 1
                     145 -> acc + 1
                     169 -> acc + 3
                     363601 -> acc + 3
                     1454 -> acc + 3
                     871 -> acc + 2
                     45361 -> acc + 2
                     872 -> acc + 2
                     45362 -> acc + 2
                     40585 -> acc + 1
                     otherwise -> go (acc + 1) $ digitFactorialSum n

chains :: [Int]
chains = [ n | n <- [1..1000000], chainLength n == 60 ]

main :: IO ()
main = print $ length chains

File: c/74_Digit_factorial_chains.c
#include <stdio.h>

int factorial[] = {1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880};

int digit_factorial_sum(int n)
{
  int sum = 0;
  for (int r = n % 10; n > 0; n /= 10, r = n % 10) {
    sum += factorial[r];
  }
  return sum;
}

int chain_len(int n)
{
  for (int len = 0; ; len += 1, n = digit_factorial_sum(n)) {
    switch (n) {
    case 1:      return len + 1;
    case 2:      return len + 1;
    case 145:    return len + 1;
    case 169:    return len + 3;
    case 363601: return len + 3;
    case 1454:   return len + 3;
    case 871:    return len + 2;
    case 45361:  return len + 2;
    case 872:    return len + 2;
    case 45362:  return len + 2;
    case 40585:  return len + 1;
    }
  }
}

int main()
{
  int num_len_60_chains = 0;
  for (int n = 1; n < 1000000; ++n) {
    if (chain_len(n) == 60)
      ++num_len_60_chains;
  }
  printf("%d\n", num_len_60_chains);
  return 0;
}
